### PR TITLE
Update `WitnessField`

### PR DESCRIPTION
**Pull request by TurtlePU**
This PR updates `WitnessField` typealias by replacing `BinaryExpansion` constraint with a conversion to the "unbounded natural numbers" type which is a `EuclideanDomain`. Also some minor improvements along the way.

Branches out from #244, so maybe that one has to be merged first.

### CODE REVIEW

**Comment by TurtlePU** [comment]
Tests fail for `UInt`'s `Ord` instance and `divMod`, investigating
Most probably something to do with variable generation again

**Comment by TurtlePU** [comment]
@vlasin, all tests pass except "Conversion to Plonk constraints and back" test because in case when some of variables are `Nothing`, constraints are not equal (and don't have to be, according to the definition of a constraint)
So this is not in scope of this PR but has to be fixed in a separate PR

**Comment by TurtlePU** [comment]
@echatav, can you take a look at this PR? It is ready.

File: src/ZkFold/Base/Algebra/Basic/Class.hs [hunk 1]
```diff
@@ -279,9 +279,10 @@ If @a@ and @b@ are in @R@ and @b@ is nonzero, then there exist @q@ and @r@ in @R
 The function @divMod@ associated with this class produces @q@ and @r@ given @a@ and @b@.
 -}
 class Semiring a => EuclideanDomain a where
```
Comment (echatav) [inline src/ZkFold/Base/Algebra/Basic/Class.hs:281]: Let's rename this to `SemiEuclidean` in this PR? `Euclidean(Domain)` should be `Ring` + `SemiEuclidean` + `eea`

**Review by echatav — CHANGES_REQUESTED** [review]
Looks good to me, but let's take the time now to rename one class?

**Review by echatav — APPROVED** [review]
:shipit:

File: src/ZkFold/Base/Algebra/Basic/Class.hs [hunk 2]
```diff
@@ -269,19 +269,24 @@ intScale n a | n < 0     = naturalFromInteger (-n) `scale` negate a
 -}
 class (AdditiveMonoid a, MultiplicativeMonoid a, FromConstant Natural a) => Semiring a
 
-{- | A Euclidean domain @R@ is an integral domain which can be endowed
-with at least one function @f : R\{0} -> R+@ s.t.
-If @a@ and @b@ are in @R@ and @b@ is nonzero, then there exist @q@ and @r@ in @R@ such that
-@a = bq + r@ and either @r = 0@ or @f(r) < f(b)@.
+{- | A semi-Euclidean-domain @a@ is a semiring without zero divisors which can
+be endowed with at least one function @f : a\{0} -> R+@ s.t. if @x@ and @y@ are
```
Comment (echatav) [inline src/ZkFold/Base/Algebra/Basic/Class.hs:273]: Screenshot 2024-09-06 at 2 47 49 PM

**Comment by vlasin** [comment]
> @vlasin, all tests pass except "Conversion to Plonk constraints and back" test because in case when some of variables are `Nothing`, constraints are not equal (and don't have to be, according to the definition of a constraint) So this is not in scope of this PR but has to be fixed in a separate PR

Could you elaborate? Why don't they have to be equal?

**Comment by TurtlePU** [comment]
> Could you elaborate? Why don't they have to be equal?

Because if one of the inputs is constant, evaluating Plonk constraint into polynomial loses information which one of them. For example,

1. The constraint `C` is `1 x c(1) + 0 x + 0 c(1) + (-1) z + 0 = 0` where `c(1)` is the placeholder constant input
2. Evaluated as a polynomial, this is `x - z = 0`
3. As a Plonk constraint, this can be represented:
  * as `C`
  * as `0 x c(1) + 1 x + 0 c(1) + (-1) z + 0 = 0`
  * as `0 c(1) x + 0 c(1) + 1 x + (-1) z + 0 = 0`
  * as `0 c(1) z + 0 c(1) + (-1) z + 1 x + 0 = 0`
  * as `0 z c(1) + (-1) z + 0 c(1) + 1 x + 0 = 0`
  * also in a myriad more ways as `a x c(1) + (1 - a) x + 0 c(1) + (-1) z + 0 = 0`
 
Before #244, the possibility of introducing `c(1)` in arbitrary Plonk constraint was negligible, so this was not caught before. Also, analogously, the same issue arises with linear constraints of the form `x + y - z = 0` because there are 6 options how this constraint can be represented, but this one again has a negligible probability in our tests because for this, we need to have the coefficient near `xy` to equal `0`.

**Comment by vlasin** [comment]
Ok, I agree to fix it in another PR.

**Pull request merged**